lessons/04-structured-output/README.md

```diff
@@ -0,0 +1,590 @@
+
+**Time beetle:** I need both of your attention, we have a problem. The aerial screw took some damage during the escape from the Roman soldiers and the landing in the Aztec Empire. Leonardo da Vinci's machine is quite sensitive and it requires very specific input to repair it. You need to provide the necessary information to your time beetle (AI Assistant) to help you repair the aerial screw.
+
+**Leonardo:** So I managed to fix the damage to the aerial screw but now it needs a boot sequence: left left up right:
+You: sounds easy enough, just that text?
+
+**Leonardo:** Yes, but it' special, it needs to be mirrored, that's how I write all my text. And of course I use a Caesar cipher with a shift of 3 because I'm Italian. :)
+
+**You:** Yes, could you just type it in then?
+
+**Leonardo:** I could, but I'm intrigued to see if you and the time beetle can do it.
+
+**You:** Really? Fine.
+
+Instruction: Write a prompt that asks for the boot sequence for the aerial screw, the text should be mirrored and encoded with a Caesar cipher with a shift of 3. Respond with the encoded text.
+
+Here's a piece of code to get started with:
+
+```javascript
+import { OpenAI } from "openai";
+
+const prompt = `TODO`; 
+
+// Call the language model with the prompt 
+
+const messages = [ 
+{ 
+    "role": "user", 
+    "content": prompt 
+}]; 
+
+// 2. Create client
+// -----------------------------------
+
+const openai = new OpenAI({
+  baseURL: "https://models.inference.ai.azure.com",
+  apiKey: process.env.GITHUB_TOKEN,
+});
+
+// 3. Send the request
+// -----------------------------------
+
+const completion = await openai.chat.completions.create({
+    model: 'gpt-4',
+    messages: messages,
+});
+
+console.log(`Answer for "${prompt}":`);
+
+// 4. Print the answer
+// -----------------------------------
+
+console.log(completion.choices[0]?.message?.content); 
+```
+
+## Solution
+
+[Solution](./solution/solution.md)
+
+## Challenge
+
+Adopt the example of trip booking and providing skills and extracting parameters to a domain of your choice. Write a prompt that asks for a user input and then extracts the necessary information to fulfill the user's request. Respond with the extracted data in JSON format with columns skill, parameters, extracted_data.
+
+Also determine the skill based on the user's input.
+
+## Summary
+
+In this chapter, we discussed structured output and how it can be used to present information in a structured manner.
+
+We explored different ways to shape the output, including instructing the model in the text, presenting an output template, and using formats like JSON, XML, or CSV.
+
+Additionally, we learned how to extract data from prompts and present it in a structured format.
+
+By leveraging structured output effectively, you can make it easier to understand and consume information generated by generative AI models.
+
+## Knowledge Check
+
+Generate a question with 2 correct responses and one incorrect:
```

Comment:
I'm not sure I understand this line - I see the question below. Are you asking them to do something else?